Feature: Ruby (phonetic guide) text - Word2007 Read/Write, HTML Read/Write, RTF write, ODT basic write

Deadpikle opened the pull request:
### Description

Right now, PHPWord does not support ruby text elements (phonetic guides) at all. You can see some documentation (of a sort) on this type of element here: https://learn.microsoft.com/en-us/dotnet/api/documentformat.openxml.wordprocessing.ruby?view=openxml-3.0.1 and some how-to-use [here](https://support.microsoft.com/en-us/office/add-phonetic-guides-to-east-asian-text-c03ad8e4-799b-499e-89a2-c749fc5397e8)

A ruby (phonetic guide) element allows you to do things like this:

<img width="557" alt="Screenshot 2025-01-23 at 8 30 35 PM" src="https://github.com/user-attachments/assets/71b737eb-274d-4c30-9200-4cf96a737ceb" />

This PR adds both reading and writing support for ruby elements to `Word2007` files. Basic text support has been checked along with Titles. Other elements may or may not be supported at this time (in other words, I didn't test literally every element with ruby); as long as the code eventually gets into `readRunChild`, the ruby should be read properly.

Please note: This is my first contribution to this repository. I may have missed something when trying to match current style guidelines, etc. Please correct me and point me in the right direction if needed. Thank you!

(Update: This PR now supports reading/writing HTML as well as writing support for RTF and basic writing support for ODT.)

### Checklist:

- [x] My CI is :green_circle:
- [x] I have covered by unit tests my new code (check build/coverage for coverage report)
- [x] I have updated the [documentation](https://github.com/PHPOffice/PHPWord/tree/master/docs) to describe the changes
- [x] I have updated the [changelog](https://github.com/PHPOffice/PHPWord/blob/master/docs/changes/1.x/1.4.0.md)

Progi1984 commented on src/PhpWord/ComplexType/RubyProperties.php:
```diff
@@ -0,0 +1,204 @@
+ *
+ * @see https://learn.microsoft.com/en-us/dotnet/api/documentformat.openxml.wordprocessing.rubyproperties?view=openxml-3.0.1
+ */
+final class RubyProperties
```
```suggestion
class RubyProperties
```

Progi1984 commented on src/PhpWord/ComplexType/RubyProperties.php:
```diff
@@ -0,0 +1,204 @@
+     *
+     * @return self
+     */
+    public function setAlignment($alignment)
```
```suggestion
    public function setAlignment(string $alignment): self
```

Progi1984 commented on src/PhpWord/ComplexType/RubyProperties.php:
```diff
@@ -0,0 +1,204 @@
+     *
+     * @return float|int
+     */
+    public function getFontFaceSize()
```
```suggestion
    public function getFontFaceSize(): float
```

Progi1984 commented on src/PhpWord/ComplexType/RubyProperties.php:
```diff
@@ -0,0 +1,204 @@
+     *
+     * @return self
+     */
+    public function setFontFaceSize($size)
```
```suggestion
    public function setFontFaceSize(float $size): self
```

Progi1984 commented on src/PhpWord/ComplexType/RubyProperties.php:
```diff
@@ -0,0 +1,204 @@
+     *
+     * @return float|int
+     */
+    public function getFontPointsAboveBaseText()
```
```suggestion
    public function getFontPointsAboveBaseText(): float
```

Progi1984 commented on src/PhpWord/ComplexType/RubyProperties.php:
```diff
@@ -0,0 +1,204 @@
+     *
+     * @return self
+     */
+    public function setFontPointsAboveBaseText($size)
```
```suggestion
    public function setFontPointsAboveBaseText(float $size): self
```

Progi1984 commented on src/PhpWord/ComplexType/RubyProperties.php:
```diff
@@ -0,0 +1,204 @@
+     *
+     * @return self
+     */
+    public function setFontSizeForBaseText($size)
```
```suggestion
    public function setFontSizeForBaseText(float $size): self
```

Progi1984 commented on src/PhpWord/ComplexType/RubyProperties.php:
```diff
@@ -0,0 +1,204 @@
+     *
+     * @return string
+     */
+    public function getLanguageId()
```
```suggestion
    public function getLanguageId(): string
```

Progi1984 commented on src/PhpWord/ComplexType/RubyProperties.php:
```diff
@@ -0,0 +1,204 @@
+     *
+     * @return self
+     */
+    public function setLanguageId($langId)
```
```suggestion
    public function setLanguageId(string $langId): self
```

Progi1984 commented on src/PhpWord/Element/Ruby.php:
```diff
@@ -0,0 +1,94 @@
+     *
+     * @return TextRun
+     */
+    public function getBaseTextRun()
```
```suggestion
    public function getBaseTextRun(): TextRun
```

Progi1984 commented on src/PhpWord/Element/Ruby.php:
```diff
@@ -0,0 +1,94 @@
+     *
+     * @return TextRun
+     */
+    public function getRubyTextRun()
```
```suggestion
    public function getRubyTextRun(): TextRun
```

Progi1984 commented on src/PhpWord/Element/Ruby.php:
```diff
@@ -0,0 +1,94 @@
+     *
+     * @return RubyProperties
+     */
+    public function getProperties()
```
```suggestion
    public function getProperties(): RubyProperties
```

Progi1984 commented on src/PhpWord/Element/Ruby.php:
```diff
@@ -0,0 +1,94 @@
+     * @param TextRun $rubyTextRun
+     * @param RubyProperties $properties
+     */
+    public function __construct($baseTextRun, $rubyTextRun, $properties)
```
```suggestion
    public function __construct(TextRun $baseTextRun, TextRun $rubyTextRun, RubyProperties $properties)
```

Progi1984 commented on src/PhpWord/Element/Ruby.php:
```diff
@@ -0,0 +1,94 @@
+ *
+ * @see https://learn.microsoft.com/en-us/dotnet/api/documentformat.openxml.wordprocessing.ruby?view=openxml-3.0.1
+ */
+class Ruby extends AbstractElement
```
Could you add setter in this class ? and unit tests about this class ?

Deadpikle replied:
Done!

Progi1984 reviewed (commented):
It's a wonderful good job.

Could you add a Sample for this new feature ?

Do you only need in Word2007 ? (not in another formats ?)

Deadpikle commented:
> Could you add a Sample for this new feature ?

I've added a new sample - `Sample_46_RubyPhoneticGuide`. I also added the getters for `Ruby` and the unit tests for that class.

> Do you only need in Word2007 ? (not in another formats ?)

I only need Word2007 writing personally, but I went ahead and added HTML writing support, too, since I might need that later and to demonstrate how that works. (Reading should be possible, I think, but looking at the code in `PhpWord/Shared/Html.php`, not sure how to do it quite right with the styles...basically, the core text in `<ruby>` goes into the base text `TextRun`, and the text inside `<rt>`, which is inside the `<ruby>` tag goes to the ruby text `TextRun`, with applicable styles along the way. That's all.)

ODT documents seem to support ruby text (see below file for sample), but the <style> is written in an interesting way, and I'm not sure how to handle that. A future PR by someone else could address ODT reading/writing. It's probably pretty straightforward at this point, but I'm not sure how to do it, so I'll let someone else tackle it.

[PhoneticGuide.odt](https://github.com/user-attachments/files/18529414/PhoneticGuide.odt)

For PDF output, `mpdf` does not support `<ruby>` tags, and I'm not sure if it has a graceful fallback or not. `dompdf` converts it to the format `{base text} ({ruby text})`, which is the default for browsers that don't support ruby out of the box (see documentation: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/rp), so dompdf at least handles it gracefully.

If someone wanted RTF output (another format I am unfamiliar with), my guess is that writing ruby text in the `{base text} ({ruby text})` format would be the way to go.

Again, I'll let someone else in another PR or worker handle those since I am not needing them and no one else has ever even asked for ruby support in this lib. :) Someone else who is more familiar with reading/writing ODT as well as writing PDF/RTF can probably handle things faster than I can at this point, anyway.

Thank you for your time and managing this lib, etc! 😄

Deadpikle commented:
I went ahead and updated the PR to add:

* HTML reading support.
* RTF writing (writes in format `base text (ruby text)`).
* Basic ODT writing (writes in format `base text (ruby text)`); this does NOT take advantage of native ODT handling at this time, and this is noted in the code.
* Added support for `TextRun::getText` when a child element is a `Ruby` element (writes `string` in format `base text (ruby text)`).

I am not planning on sneaking in native ODT support later 😅 Someone else can do that if they ever need it in the future. At least for now, ODT files now don't totally ignore ruby elements, which is an improvement to before. So this PR is ready to go.

Progi1984 commented on src/PhpWord/ComplexType/RubyProperties.php:
```diff
@@ -0,0 +1,188 @@
+ *
+ * @see https://learn.microsoft.com/en-us/dotnet/api/documentformat.openxml.wordprocessing.rubyproperties?view=openxml-3.0.1
+ */
+class RubyProperties
```
@Deadpikle It misses unit tests for RubyProperties.

Deadpikle replied:
Oops, sorry about that. I've added some unit tests for `RubyProperties`. 😄

Progi1984 reviewed (commented):
One last comment and it will be ok.

Deadpikle commented:
Tests have been added. Thanks again.

Progi1984 commented:
Thank you for your wonderful contribution, @Deadpikle  :tada:
